src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,836 @@
+            return Self::Exclude {
+                need_null,
+                databases,
+            };
+        }
+
+        Self::None
+    }
+
+    fn remove_empty(mut names: HashSet<String>) -> (bool, Vec<String>) {
+        let null = names.remove(&"".to_owned());
+        (null, names.into_iter().collect())
+    }
+
+    async fn pg_stat_database(&self, client: &Client) -> Result<Vec<Row>, PgError> {
+        let mut conditions = "SELECT * FROM pg_stat_database".to_owned();
+        match self {
+            Self::Include {
+                need_null,
+                databases,
+            } => {
+                conditions += " WHERE datname = ANY($1)";
+                if *need_null {
+                    conditions += " OR datname IS NULL";
+                }
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::Exclude {
+                need_null,
+                databases,
+            } => {
+                conditions += " WHERE datname != ANY($1)";
+                if *need_null {
+                    conditions += " AND datname IS NOT NULL";
+                } else {
+                    conditions += " OR datname IS NULL";
+                }
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::None => client.query(conditions.as_str(), &[]).await,
+        }
+    }
+
+    async fn pg_stat_database_conflicts(&self, client: &Client) -> Result<Vec<Row>, PgError> {
+        let mut conditions = "SELECT * FROM pg_stat_database_conflicts".to_owned();
+        match self {
+            Self::Include { databases, .. } => {
+                conditions += " WHERE datname = ANY($1)";
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::Exclude { databases, .. } => {
+                conditions += " WHERE datname != ANY($1)";
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::None => client.query(conditions.as_str(), &[]).await,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct PostgresqlMetrics {
+    config: Config,
+    tls_config: Option<PostgresqlMetricsTlsConfig>,
+    client: Option<Client>,
+    namespace: Option<String>,
+    tags: BTreeMap<String, String>,
+    datname_filter: DatnameFilter,
+}
+
+impl PostgresqlMetrics {
+    async fn new(
+        endpoint: String,
+        datname_filter: DatnameFilter,
+        namespace: Option<String>,
+        tls_config: Option<PostgresqlMetricsTlsConfig>,
+    ) -> Result<Self, BuildError> {
+        let config: Config = endpoint.parse().context(InvalidEndpoint)?;
+        if config.get_hosts().len() > 1 {
```

Comment:
Should we check for `config.get_hosts().len() == 0` too?

Reply:
Yes, we need, otherwise, we will get `index out of bound` on the next steps =\